eknrenderer/eknr-renderer.c

```diff
@@ -0,0 +1,1306 @@
+                            const char     *buffer,
+                            uintmax_t       buffer_size)
+{
+  RendererMustacheData *data = userdata;
+  return mustache_std_strwrite (api, (void *) &data->output, (char *) buffer, buffer_size);
+}
+
+static const char *
+_lookup_in_gvariant_dict (RendererMustacheData *data,
+                          const char           *text,
+                          mustache_api_t       *api)
+{
+  g_autoptr(GVariant) value_v = g_variant_dict_lookup_value (data->variables, text, G_VARIANT_TYPE_STRING);
+
+  if (value_v == NULL)
+    {
+      g_autofree char *msg = g_strdup_printf ("No such variable %s", text);
+      (*api->error) (api, data, __LINE__, msg);
+      return NULL;
+    }
+
+  return g_variant_get_string (value_v, NULL);
+}
+
+static const char *
+html_entity_for_char (char matched_ent)
+{
+  switch (matched_ent) {
+    case '&':
+      return "&amp;";
+    case '<':
+      return "&lt;";
+    case '>':
+      return "&gt;";
+    case '"':
+      return "&quot;";
+    case '\'':
+      return "&#39;";
+    case '/':
+      return "&#x2F;";
+    default:
+      return NULL;
+  }
+}
+
+static gboolean
+escape_value_eval_callback (const GMatchInfo       *match_info,
+                            GString                *result,
+                            G_GNUC_UNUSED gpointer  user_data)
+{
+  g_autofree char *matched = g_match_info_fetch (match_info, 0);
+  char matched_ent = matched[0];
+  const char *html_entity = html_entity_for_char (matched_ent);
+
+  if (html_entity == NULL)
+    return TRUE;
+
+  g_string_append (result, html_entity);
+
+  /* From the documentation, returning FALSE to continues the replacement
+   * process, returning TRUE stops it. It isn't entirely clear, but it seems
+   * that prematurely stopping the replacement process is an error case. */
+  return FALSE;
+}
+
+/**
+ * eknr_escape_html:
+ * @html: The string to escape
+ * @error: A #GError
+ *
+ * Escape HTML, replacing characters with HTML entities.
+ *
+ * Returns: (transfer full): The escaped HTML document, or %NULL on error.
+ */
+char *
+eknr_escape_html (const char *html,
+                  GError    **error)
+{
+  g_autoptr(GRegex) regex = NULL;
+  g_autofree char *escaped = NULL;
+
+  /* First, use a regex to pick out the relevant entities that
+   * need to be escaped.*/
+  regex = g_regex_new ("[&<>\"'\\/]", 0, 0, error);
+
+  if (regex == NULL)
+    return NULL;;
+
+  escaped = g_regex_replace_eval (regex,
+                                  html,
+                                  strlen (html),
+                                  0,
+                                  0,
+                                  escape_value_eval_callback,
+                                  NULL,
+                                  error);
+
+  if (escaped == NULL)
+    return NULL;
+
+  return g_steal_pointer (&escaped);
+}
+
+/**
+ * maybe_escape_value:
+ * @value: (transfer full): The string to escape
```

Comment:
This is not `(transfer full)`. I'd remove this gtk-doc comment here, maybe just keep the body of this comment in the function itself.

Reply:
Ah, yes, this is no longer `(transfer full)` - was a leftover